### ISSUE TITLE
Calculating ROIs based off of labels

### ISSUE THREAD

**Issue by ajc3xc** [danforthcenter/plantcv#1509]
When I was working on my plant phenotyping Capstone project for MarsFarm, when I was trying to calculate rois automatically for a diverse set of images I developed a program to automatically calculate roi points and roi sizes based off of plant labeled masks. I used the midpoint of each label as the points of the rois, and I used a special algorithm to calculate roi size. Below is a repository containing an example running this program, where calculate_rois.py is the program used for calculating the rois. Here is the [repository](https://github.com/MST-Capstone2-MarsFarm/Tutorials/blob/auto_roi_example).

**Comment by HaleySchuhl** [comment]
Hi @ajc3xc thank you very much for opening this issue and contributing this new feature. It seems helpful! Are you interested in drafting a Pull Request yourself or would you prefer someone on our team take the lead on the PR? 

Also, I ran into an error while testing on the example image provided in your repository. I haven't tried troubleshooting this error yet. Have you been able to resolve this in your own environment? 

```
TypeError                                 Traceback (most recent call last)
Cell In[2], line 42
     39 labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)
     41 #get the centers and optimal radius size
---> 42 centers, optimal_radius_size = auto_roi(labeled_image)
     44 #now that this is calculated, use pcv.roi.multi to automatically generate the rois
     45 rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)

TypeError: cannot unpack non-iterable auto_roi object
```

- the workflow I'm using 
```
from plantcv import plantcv as pcv
import cv2
import numpy as np
from skimage import morphology
from calculate_rois import auto_roi

image_path = "plant_image.jpg"

img, _, _ = pcv.readimage(image_path)

#mask the image with an example mask. Filter out small areas
# Convert to HSV and LAB color spaces
hsv_image = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
lab_image = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)

# Define HSV range for filtering using OpenCV
# OpenCV uses 0-180 for Hue, so the values are halved
hsv_min = np.array([int(28/2), int(20/100*255), int(20/100*255)])
hsv_max = np.array([int(144/2), 255, 255])
hsv_mask = cv2.inRange(hsv_image, hsv_min, hsv_max)

# Define LAB range for filtering using OpenCV
# OpenCV uses 0-255 for L, a*, and b*
# Note: 'a' and 'b' ranges need to be shifted from [-128, 127] to [0, 255]
# L is scaled from [0, 100] in LAB to [0, 255] in OpenCV
lab_lower = np.array([int(10/100*255), 0, 132])
lab_upper = np.array([int(90/100*255), 124, 255])
lab_mask = cv2.inRange(lab_image, lab_lower, lab_upper)

# Combine the masks (logical AND) and apply to the original image
combined_mask = cv2.bitwise_and(hsv_mask, lab_mask)

#remove small holes in the image so object detection won't be as bad
# Remove small objects
cleaned_mask = morphology.remove_small_objects(combined_mask, min_size=300)

#generate automatic labels from denoised mask
#create labels for masks that may have plants, count number of objects
labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)

#get the centers and optimal radius size
centers, optimal_radius_size = auto_roi(labeled_image)

#now that this is calculated, use pcv.roi.multi to automatically generate the rois
rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)from plantcv import plantcv as pcv
import cv2
import numpy as np
from skimage import morphology
from calculate_rois import auto_roi

image_path = "plant_image.jpg"

img, _, _ = pcv.readimage(image_path)

#mask the image with an example mask. Filter out small areas
# Convert to HSV and LAB color spaces
hsv_image = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
lab_image = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)

# Define HSV range for filtering using OpenCV
# OpenCV uses 0-180 for Hue, so the values are halved
hsv_min = np.array([int(28/2), int(20/100*255), int(20/100*255)])
hsv_max = np.array([int(144/2), 255, 255])
hsv_mask = cv2.inRange(hsv_image, hsv_min, hsv_max)

# Define LAB range for filtering using OpenCV
# OpenCV uses 0-255 for L, a*, and b*
# Note: 'a' and 'b' ranges need to be shifted from [-128, 127] to [0, 255]
# L is scaled from [0, 100] in LAB to [0, 255] in OpenCV
lab_lower = np.array([int(10/100*255), 0, 132])
lab_upper = np.array([int(90/100*255), 124, 255])
lab_mask = cv2.inRange(lab_image, lab_lower, lab_upper)

# Combine the masks (logical AND) and apply to the original image
combined_mask = cv2.bitwise_and(hsv_mask, lab_mask)

#remove small holes in the image so object detection won't be as bad
# Remove small objects
cleaned_mask = morphology.remove_small_objects(combined_mask, min_size=300)

#generate automatic labels from denoised mask
#create labels for masks that may have plants, count number of objects
labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)

#get the centers and optimal radius size
centers, optimal_radius_size = auto_roi(labeled_image)

#now that this is calculated, use pcv.roi.multi to automatically generate the rois
rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)
```

**Comment by ajc3xc** [comment]
> Hi @ajc3xc thank you very much for opening this issue and contributing this new feature. It seems helpful! Are you interested in drafting a Pull Request yourself or would you prefer someone on our team take the lead on the PR?
> 
> Also, I ran into an error while testing on the example image provided in your repository. I haven't tried troubleshooting this error yet. Have you been able to resolve this in your own environment?
> 
> ```
> TypeError                                 Traceback (most recent call last)
> Cell In[2], line 42
>      39 labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)
>      41 #get the centers and optimal radius size
> ---> 42 centers, optimal_radius_size = auto_roi(labeled_image)
>      44 #now that this is calculated, use pcv.roi.multi to automatically generate the rois
>      45 rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)
> 
> TypeError: cannot unpack non-iterable auto_roi object
> ```
> 
> * the workflow I'm using
> 
> ```
> from plantcv import plantcv as pcv
> import cv2
> import numpy as np
> from skimage import morphology
> from calculate_rois import auto_roi
> 
> image_path = "plant_image.jpg"
> 
> img, _, _ = pcv.readimage(image_path)
> 
> #mask the image with an example mask. Filter out small areas
> # Convert to HSV and LAB color spaces
> hsv_image = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
> lab_image = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)
> 
> # Define HSV range for filtering using OpenCV
> # OpenCV uses 0-180 for Hue, so the values are halved
> hsv_min = np.array([int(28/2), int(20/100*255), int(20/100*255)])
> hsv_max = np.array([int(144/2), 255, 255])
> hsv_mask = cv2.inRange(hsv_image, hsv_min, hsv_max)
> 
> # Define LAB range for filtering using OpenCV
> # OpenCV uses 0-255 for L, a*, and b*
> # Note: 'a' and 'b' ranges need to be shifted from [-128, 127] to [0, 255]
> # L is scaled from [0, 100] in LAB to [0, 255] in OpenCV
> lab_lower = np.array([int(10/100*255), 0, 132])
> lab_upper = np.array([int(90/100*255), 124, 255])
> lab_mask = cv2.inRange(lab_image, lab_lower, lab_upper)
> 
> # Combine the masks (logical AND) and apply to the original image
> combined_mask = cv2.bitwise_and(hsv_mask, lab_mask)
> 
> #remove small holes in the image so object detection won't be as bad
> # Remove small objects
> cleaned_mask = morphology.remove_small_objects(combined_mask, min_size=300)
> 
> #generate automatic labels from denoised mask
> #create labels for masks that may have plants, count number of objects
> labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)
> 
> #get the centers and optimal radius size
> centers, optimal_radius_size = auto_roi(labeled_image)
> 
> #now that this is calculated, use pcv.roi.multi to automatically generate the rois
> rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)from plantcv import plantcv as pcv
> import cv2
> import numpy as np
> from skimage import morphology
> from calculate_rois import auto_roi
> 
> image_path = "plant_image.jpg"
> 
> img, _, _ = pcv.readimage(image_path)
> 
> #mask the image with an example mask. Filter out small areas
> # Convert to HSV and LAB color spaces
> hsv_image = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
> lab_image = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)
> 
> # Define HSV range for filtering using OpenCV
> # OpenCV uses 0-180 for Hue, so the values are halved
> hsv_min = np.array([int(28/2), int(20/100*255), int(20/100*255)])
> hsv_max = np.array([int(144/2), 255, 255])
> hsv_mask = cv2.inRange(hsv_image, hsv_min, hsv_max)
> 
> # Define LAB range for filtering using OpenCV
> # OpenCV uses 0-255 for L, a*, and b*
> # Note: 'a' and 'b' ranges need to be shifted from [-128, 127] to [0, 255]
> # L is scaled from [0, 100] in LAB to [0, 255] in OpenCV
> lab_lower = np.array([int(10/100*255), 0, 132])
> lab_upper = np.array([int(90/100*255), 124, 255])
> lab_mask = cv2.inRange(lab_image, lab_lower, lab_upper)
> 
> # Combine the masks (logical AND) and apply to the original image
> combined_mask = cv2.bitwise_and(hsv_mask, lab_mask)
> 
> #remove small holes in the image so object detection won't be as bad
> # Remove small objects
> cleaned_mask = morphology.remove_small_objects(combined_mask, min_size=300)
> 
> #generate automatic labels from denoised mask
> #create labels for masks that may have plants, count number of objects
> labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)
> 
> #get the centers and optimal radius size
> centers, optimal_radius_size = auto_roi(labeled_image)
> 
> #now that this is calculated, use pcv.roi.multi to automatically generate the rois
> rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)
> ```

The bug was due to an error in the code. I made some changes to the code to address the issue. I also added a jupyter notebook to make a visual example of its functionality with an example image.

I don't have much experience making a pull request. Would I want to base my changes off of main?  Would there be a specific group of functions that this class would go under? I would be more than fine doing it, I just don't know too much about how they work. Thanks in advance for any information.

**Comment by HaleySchuhl** [comment]
Hi @ajc3xc apologies for the delay in response. This is great, I'm happy to help you get this code into a Pull Request. Yes, you'll want to make a new branch based on the main branch, and the location for the new code can go into `plantcv.plantcv.roi.roi_methods.py`. I haven't looked too deeply at your source code yet but I'm curious if you think the function could be refactored without a class or whether that structure is necessary for this tool. We've got a doc page about [contributing](https://plantcv.readthedocs.io/en/latest/CONTRIBUTING/) but please don't hesitate to ask questions here that we don't cover in documentation. 

I'll also make sure @nfahlgren has added you as a contributor to the repository so you can open a branched PR rather than a fork. Thanks in advance!

**Comment by ajc3xc** [comment]
@HaleySchuhl I wasn't able to add a new branch so I made a fork to add my code into ```plantcv.plantcv.roi.roi_methods.py```. I converted the class into a function so that it would be consistent with the rest of the code. There are subfunctions within the main function, so it is a bit hard to follow. I have tested out the code and it seemed to work. [Here is the fork](https://github.com/ajc3xc/plantcv_multiroi).

Also, the location and example image where I tested out this code on one image is [here](https://github.com/MST-Capstone2-MarsFarm/Tutorials/tree/auto_roi_example_function)

**Comment by HaleySchuhl** [comment]
This is fantastic @ajc3xc thank you. Since this message I have reached out to Noah and you should now have permissions to branch and copy those changes in `plantcv.plantcv.roi.roi_methods.py` over. Then you can publish the branch here to GitHub, and from there, I'm happy to help you with the other components of a pull request (namely the init files, adding unit test(s), adding a documentation page and including the new doc page in the table of contents `mkdocs.yml`). This is outlined [here](https://plantcv.readthedocs.io/en/latest/CONTRIBUTING/#create-a-branch-to-do-your-work) but I'm also happy to setup a quick zoom call to talk through the process, just let me know if this would be helpful.

**Comment by ajc3xc** [comment]
Ok, the changes should be under the branch [label_rois](https://github.com/danforthcenter/plantcv/tree/labels_rois). Glad to contribute to this project!

**Comment by HaleySchuhl** [comment]
@ajc3xc this is great thank you! I went ahead and just opened a Pull Request from your branch, and I will pull down the changes that were merged into PlantCV this past Friday. I aim to review the new function tomorrow and will ping you with any comments.

**Comment by HaleySchuhl** [comment]
Happy December @ajc3xc and apologies for the long delay. I was just testing the following version of the workflow on the test image linked here, but I'm struggling to replicate the functionality. I'm getting results where the optimal radius size is calculated to be 0 somehow? 

Taking a step back, my understanding is that this function aims to separate touching objects by auto-calculating ROIs that can be later used to `cutto` in the mask, does that sound accurate?  It seems like a useful tool, but I haven't been able to get it to work for me locally still. 

*Workflow version I'm testing* 
```
from plantcv import plantcv as pcv
import cv2
import numpy as np
from skimage import morphology
from calculate_rois import auto_roi
pcv.params.debug = "plot" 
image_path = "plant_image.jpg"

img, _, _ = pcv.readimage(image_path)

#mask the image with an example mask. Filter out small areas
# Convert to HSV and LAB color spaces
hsv_image = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
lab_image = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)

# Define HSV range for filtering using OpenCV
# OpenCV uses 0-180 for Hue, so the values are halved
hsv_min = np.array([int(28/2), int(20/100*255), int(20/100*255)])
hsv_max = np.array([int(144/2), 255, 255])
hsv_mask = cv2.inRange(hsv_image, hsv_min, hsv_max)

# Define LAB range for filtering using OpenCV
# OpenCV uses 0-255 for L, a*, and b*
# Note: 'a' and 'b' ranges need to be shifted from [-128, 127] to [0, 255]
# L is scaled from [0, 100] in LAB to [0, 255] in OpenCV
lab_lower = np.array([int(10/100*255), 0, 132])
lab_upper = np.array([int(90/100*255), 124, 255])
lab_mask = cv2.inRange(lab_image, lab_lower, lab_upper)

# Combine the masks (logical AND) and apply to the original image
combined_mask = cv2.bitwise_and(hsv_mask, lab_mask)

#remove small holes in the image so object detection won't be as bad
# Remove small objects
cleaned_mask = morphology.remove_small_objects(combined_mask, min_size=300)

#generate automatic labels from denoised mask
#create labels for masks that may have plants, count number of objects
labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)
#get the centers and optimal radius size
centers, optimal_radius_size = auto_roi(labeled_image)
pcv.params.debug = "plot" 
#now that this is calculated, use pcv.roi.multi to automatically generate the rois
rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)
```

*Function version I'm testing* 
```
import math
import numpy as np 

def auto_roi(labeled_objects_array: np.ndarray=None) -> list[list[tuple[int, int]], int]:
    labeled_objects_array = labeled_objects_array
    #the number of unique plants should all the unique np values from the array that aren't zero
    #since 0 is counted as a number, should be len - 1
    num_plants = len(np.unique(labeled_objects_array)) - 1
    center_points = [] 
    
   #calculate the euclidean distance between two points

    #the 'main' function of the class
    def get_roi_centers_and_individual_images():        
        def _euclidian_distance(point1: tuple, point2: tuple):
            """Calculate the Euclidean distance between two points."""
            return math.sqrt((point1[0] - point2[0])**2 + (point1[1] - point2[1])**2)
    
        #get minimum of either distance to edge or distance to another plant
        def _min_no_intersect_euclidean_distance():
            #get minimum distance of any of the points to an edge
            def _min_distance_to_edge():
                # x is 0th index, y is 1th index in tuple (x, y)
                height, width = np.shape(labeled_objects_array)[:2]
                x_vals = [x for x, y in center_points]
                y_vals = [y for x, y in center_points]
                xmin, xmax = min(x_vals), max(x_vals)
                ymin, ymax = min(y_vals), max(y_vals)
                distances_to_edge = [xmin, width-xmax, ymin, height-ymax]
                return min(distances_to_edge)
            
            """Find the maximum Euclidean distance between any two points in the list."""
            if len(center_points) <= 1: return 0
            n = len(center_points)
        
            min_distance = _min_distance_to_edge()
        
            #calculate euclidean distances between points and to the edge of the graph
            for i in range(1, n):
                for j in range(i + 1, n):
                    distance = _euclidian_distance(center_points[i], center_points[j])
                    min_distance = min(min_distance, distance)
            return math.floor(min_distance)

        #no lower bound for the radius
        #this should help accomadate identifying small individual sprouts
        max_circle_radius_size = 0
        
        for number in range(1, num_plants+1):
            # Create a mask for the current number
            mask = labeled_objects_array == number
            
            # Find indices where the mask is True
            cols, rows = np.nonzero(mask)
            
            if rows.size > 0 and cols.size > 0:
                # Calculate bounds
                #rows = y axis, columns = x axis, depth = z axis
                min_row, max_row = np.min(rows), np.max(rows)
                min_col, max_col = np.min(cols), np.max(cols)
    
                middle_row = (min_row + max_row) // 2
                middle_col = (min_col + max_col) // 2

                #calculate the radius of a circle that would fit the entire plant
                #radius = diameter / 2
                point1 = (min_row, min_col)
                point2 = (max_row, max_col)
                encapsulating_circle_radius = _euclidian_distance(point1, point2) // 2

                #once that's done, then iteratively calculate the max radius size for
                max_circle_radius_size = max(max_circle_radius_size, encapsulating_circle_radius)
    
                #calculate the radius 
    
                center_points.append((middle_row, middle_col))                                                                        
                
            else:
                print(f"no indices for {number}")
                continue
        #calculate minimum distance either to edge, distance to another plant, or (the largest radius or any other plant with a floor of 50)
        optimal_radius_size = min(_min_no_intersect_euclidean_distance(), encapsulating_circle_radius)
        
        return center_points, optimal_radius_size
    
    return get_roi_centers_and_individual_images()
```

* Results* 

![image](https://github.com/user-attachments/assets/b7bbc8c3-1bc7-4e6c-add2-46fd38c0df3e)

**Comment by ajc3xc** [comment]
Hi Haley,

The issue should be in the auto_roi_example not properly removing small objects. This should fix the issue:

boolean_mask = combined_mask > 0

#remove small holes in the image so object detection won't be as bad
# Remove small objects
cleaned_mask = morphology.remove_small_objects(boolean_mask, min_size=1000)
cleaned_mask = cleaned_mask.astype(np.uint8) * 255

**Comment by HaleySchuhl** [comment]
Hi Adam, thanks for providing this update. I'm thinking I'm still not getting the intended result. 

Here's the latest workflow I tried: 

```
from plantcv import plantcv as pcv
import cv2
import numpy as np
from skimage import morphology
from calculate_rois import auto_roi
pcv.params.debug = "plot" 
image_path = "plant_image.jpg"

img, _, _ = pcv.readimage(image_path)

#mask the image with an example mask. Filter out small areas
# Convert to HSV and LAB color spaces
hsv_image = cv2.cvtColor(img, cv2.COLOR_BGR2HSV)
lab_image = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)

# Define HSV range for filtering using OpenCV
# OpenCV uses 0-180 for Hue, so the values are halved
hsv_min = np.array([int(28/2), int(20/100*255), int(20/100*255)])
hsv_max = np.array([int(144/2), 255, 255])
hsv_mask = cv2.inRange(hsv_image, hsv_min, hsv_max)

# Define LAB range for filtering using OpenCV
# OpenCV uses 0-255 for L, a*, and b*
# Note: 'a' and 'b' ranges need to be shifted from [-128, 127] to [0, 255]
# L is scaled from [0, 100] in LAB to [0, 255] in OpenCV
lab_lower = np.array([int(10/100*255), 0, 132])
lab_upper = np.array([int(90/100*255), 124, 255])
lab_mask = cv2.inRange(lab_image, lab_lower, lab_upper)

# Combine the masks (logical AND) and apply to the original image
combined_mask = cv2.bitwise_and(hsv_mask, lab_mask)

boolean_mask = combined_mask > 0

#remove small holes in the image so object detection won't be as bad
# Remove small objects
cleaned_mask = morphology.remove_small_objects(boolean_mask, min_size=1000)
cleaned_mask = cleaned_mask.astype(np.uint8) * 255


#generate automatic labels from denoised mask
#create labels for masks that may have plants, count number of objects
labeled_image, number_of_plants = pcv.create_labels(mask=cleaned_mask)

#get the centers and optimal radius size
centers, optimal_radius_size = auto_roi(labeled_image)
pcv.params.debug = "plot" 
#now that this is calculated, use pcv.roi.multi to automatically generate the rois
rois = pcv.roi.multi(img=img, coord=centers, radius=optimal_radius_size)
```

Here's the version of the function I'm using (the function within a function syntax needs reformatting still):

```
import math
import numpy as np 

def auto_roi(labeled_objects_array: np.ndarray=None) -> list[list[tuple[int, int]], int]:
    labeled_objects_array = labeled_objects_array
    #the number of unique plants should all the unique np values from the array that aren't zero
    #since 0 is counted as a number, should be len - 1
    num_plants = len(np.unique(labeled_objects_array)) - 1
    center_points = [] 
    
   #calculate the euclidean distance between two points

    #the 'main' function of the class
    def get_roi_centers_and_individual_images():        
        def _euclidian_distance(point1: tuple, point2: tuple):
            """Calculate the Euclidean distance between two points."""
            return math.sqrt((point1[0] - point2[0])**2 + (point1[1] - point2[1])**2)
    
        #get minimum of either distance to edge or distance to another plant
        def _min_no_intersect_euclidean_distance():
            #get minimum distance of any of the points to an edge
            def _min_distance_to_edge():
                # x is 0th index, y is 1th index in tuple (x, y)
                height, width = np.shape(labeled_objects_array)[:2]
                x_vals = [x for x, y in center_points]
                y_vals = [y for x, y in center_points]
                xmin, xmax = min(x_vals), max(x_vals)
                ymin, ymax = min(y_vals), max(y_vals)
                distances_to_edge = [xmin, width-xmax, ymin, height-ymax]
                return min(distances_to_edge)
            
            """Find the maximum Euclidean distance between any two points in the list."""
            if len(center_points) <= 1: return 0
            n = len(center_points)
        
            min_distance = _min_distance_to_edge()
        
            #calculate euclidean distances between points and to the edge of the graph
            for i in range(1, n):
                for j in range(i + 1, n):
                    distance = _euclidian_distance(center_points[i], center_points[j])
                    min_distance = min(min_distance, distance)
            return math.floor(min_distance)

        #no lower bound for the radius
        #this should help accomadate identifying small individual sprouts
        max_circle_radius_size = 0
        
        for number in range(1, num_plants+1):
            # Create a mask for the current number
            mask = labeled_objects_array == number
            
            # Find indices where the mask is True
            cols, rows = np.nonzero(mask)
            
            if rows.size > 0 and cols.size > 0:
                # Calculate bounds
                #rows = y axis, columns = x axis, depth = z axis
                min_row, max_row = np.min(rows), np.max(rows)
                min_col, max_col = np.min(cols), np.max(cols)
    
                middle_row = (min_row + max_row) // 2
                middle_col = (min_col + max_col) // 2

                #calculate the radius of a circle that would fit the entire plant
                #radius = diameter / 2
                point1 = (min_row, min_col)
                point2 = (max_row, max_col)
                encapsulating_circle_radius = _euclidian_distance(point1, point2) // 2

                #once that's done, then iteratively calculate the max radius size for
                max_circle_radius_size = max(max_circle_radius_size, encapsulating_circle_radius)
    
                #calculate the radius 
    
                center_points.append((middle_row, middle_col))                                                                        
                
            else:
                print(f"no indices for {number}")
                continue
        #calculate minimum distance either to edge, distance to another plant, or (the largest radius or any other plant with a floor of 50)
        optimal_radius_size = min(_min_no_intersect_euclidean_distance(), encapsulating_circle_radius)
        
        return center_points, optimal_radius_size
    
    return get_roi_centers_and_individual_images()
```

**Resulting ROIS**

![Image](https://github.com/user-attachments/assets/099cf952-d936-4ef6-ac23-a2957809881b)
